// express-cassandra/lib/helpers/driver.js
'use strict';

var _ = require('lodash');

var debug = require('debug')('express-cassandra');

var Driver = function f(properties) {
  this._properties = properties;
};

Driver.prototype = {
  ensure_init(callback) {
    if (!this._properties.cql) {
      this._properties.init(callback);
    } else {
      callback();
    }
  },

  execute_definition_query(query, params, callback) {
    var _this = this;

    this.ensure_init(function (err) {
      if (err) {
        callback(err);
        return;
      }
      debug('executing definition query: %s with params: %j', query, params);
      var properties = _this._properties;
      var conn = properties.define_connection;
      conn.execute(query, params, { prepare: false, fetchSize: 0 }, callback);
    });
  },

  execute_query(query, params, options, callback) {
    var _this2 = this;

    if (arguments.length === 3) {
      callback = options;
      options = {};
    }

    var defaults = {
      prepare: true
    };

    options = _.defaultsDeep(options, defaults);

    this.ensure_init(function (err) {
      if (err) {
        callback(err);
        return;
      }
      debug('executing query: %s with params: %j', query, params);
      _this2._properties.cql.execute(query, params, options, function (err1, result) {
        if (err1 && err1.code === 8704) {
          _this2.execute_definition_query(query, params, callback);
        } else {
          callback(err1, result);
        }
      });
    });
  },

  execute_batch(queries, options, callback) {
    var _this3 = this;

    if (arguments.length === 2) {
      callback = options;
      options = {};
    }

    var defaults = {
      prepare: true
    };

    options = _.defaultsDeep(options, defaults);

    this.ensure_init(function (err) {
      if (err) {
        callback(err);
        return;
      }
      debug('executing batch queries: %j', queries);
      _this3._properties.cql.batch(queries, options, callback);
    });
  },

  execute_eachRow(query, params, options, onReadable, callback) {
    var _this4 = this;

    this.ensure_init(function (err) {
      if (err) {
        callback(err);
        return;
      }
      debug('executing eachRow query: %s with params: %j', query, params);
      _this4._properties.cql.eachRow(query, params, options, onReadable, callback);
    });
  },

  execute_stream(query, params, options, onReadable, callback) {
    var _this5 = this;

    this.ensure_init(function (err) {
      if (err) {
        callback(err);
        return;
      }
      debug('executing stream query: %s with params: %j', query, params);
      _this5._properties.cql.stream(query, params, options).on('readable', onReadable).on('end', callback);
    });
  }
};

module.exports = Driver;
